iotile_analytics_interactive/iotile_analytics/interactive/reports/cloud_uploader.py

```diff
@@ -0,0 +1,232 @@
+"""Helper class that manages uploading a series of report files to iotile.cloud."""
+
+from __future__ import unicode_literals, print_function, absolute_import
+import logging
+import os
+import json
+import string
+import random
+from iotile_analytics.core import CloudSession
+from iotile_analytics.core.exceptions import UsageError, CloudError
+from typedargs.exceptions import ArgumentError
+
+
+class ReportUploader(object):
+    """Upload a series of files to S3 using signed URLs.
+
+    Args:
+        domain (str): The iotile.cloud instance that we should use. If
+            not passed, this defaults to https://iotile.cloud
+    """
+
+    def __init__(self, domain=None):
+        self._session = CloudSession(domain=domain)
+        self._api = self._session._api
+        self._logger = logging.getLogger(__name__)
+
+    def upload_report(self, label, files, group=None, slug=None, org=None):
+        """Upload a report to iotile.cloud to attach to a device or archive.
+
+        The report will be attached to the given AnalysisGroup, or you can
+        specify the slug of the object (and its org) to attach the report to
+        explicitly.  You must specify EITHER group OR (slug and org) and the
+        currently logged in user must have permission to upload reports for
+        the given org.
+
+        You must give the report a human readable label that will be shown
+        in lists where users may select to view the report.
+
+        Args:
+            label (str): The human readable label for the report that you
+                are uploading.
+            files (list of str): A list of all of the files that should
+                be uploaded for this report including the main entrypoint
+                file that should be listed first.
+            group (AnalysisGroup): The analysis group that was used to generate
+                this report.  If this is not specified then you must explicitly
+                specify the slug of the object used to generate the report and
+                the slug of the organization that it should be uploaded for.
+
+
+        Returns:
+            str: A URL where the uploaded report can be accessed.
+        """
+
+        if group is not None:
+            slug = group.source_info.get('slug')
+            org = group.source_info.get('org')
+
+            if slug is None or org is None:
+                raise ArgumentError("The group provided did not have org or source slug information in its source_info", slug=slug, org=org)
+        elif org is None or slug is None:
+            raise UsageError("If you do not specify an AnalysisGroup you must pass an org and a slug to this function", org=org, slug=slug)
+
+        report_id = self._create_report(label, slug, org)
+        self._logger.debug("Created report id: %s", report_id)
+
+        keys = self._clean_file_paths(files)
+
+        urls, fields = self._get_signed_urls(report_id, keys)
```

Comment:
It seems you first get all signed URLs, and then upload them.
Any reason why you don't do it as you go? i.e. get url and upload.
The reason I am asking is that all the signed URLs have an expiration time. I don't think this is an issue for this application but in theory there could be a situation where the URL expires before the upload happens

Reply:
Interesting.  We can refactor this if that becomes an issue.  The reason is that the parallelization for the requests is handled by `CloudSession` and there wasn't a good way to express 'post this' then get the result, call this function on it and then post again, whereas I could just make a `post_multiple` function that did it in stages but I see the point.

Right now getting the signatures is fast, just a few seconds for 300 of them.